server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -669,56 +683,177 @@ public String toString() {
+        protected void doRun() throws Exception {
+            assert ThreadPool.assertCurrentMethodIsNotCalledRecursively();
+            assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[send file chunk]");
+            while (true) {
+                assert semaphore.availablePermits() == 0;
+                if (error.get() != null) {
+                    handleErrorOnSendFiles(store, error.get().v2(), new StoreFileMetaData[]{error.get().v1()});
+                    throw error.get().v2();
+                }
+                final FileChunk chunk = readNextChunk();
+                if (chunk == null) {
+                    semaphore.release(); // allow other threads respond if we are not done yet.
+                    if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getProcessedCheckpoint() && semaphore.tryAcquire()) {
+                        listener.onResponse(null);
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.getAsInt(),
-                            ActionListener.wrap(
-                                r -> requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId),
-                                e -> {
-                                    error.compareAndSet(null, Tuple.tuple(md, e));
-                                    requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId);
+                    break;
+                }
+                final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                cancellableThreads.execute(() ->
+                    recoveryTarget.writeFileChunk(chunk.md, chunk.position, chunk.content, chunk.lastChunk, translogOps.getAsInt(),
+                        ActionListener.wrap(
+                            r -> {
+                                chunk.close(); // release the buffer so we can reuse to reduce allocation
+                                requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId);
+                                if (canSendMore() && semaphore.tryAcquire()) {
+                                    sendFileExecutor.execute(this); // fork off from the network thread
+                                }
+                            },
+                            e -> {
+                                if (error.compareAndSet(null, Tuple.tuple(chunk.md, e)) && semaphore.tryAcquire()) {
+                                    // have to fork as handleErrorOnSendFiles can read file which should not happen on the network thread.
+                                    sendFileExecutor.execute(this);
                                 }
-                            )));
-                    position += content.length();
+                            })
+                    )
+                );
+                if (canSendMore() == false) {
```

Comment:
Maybe it was easier to try-take the lock as first thing in while loop and bail out (return) if not taken. Then check `canSendMore()` afterwards. If my previous comment holds, we should only check `canSendMore()` under the lock/semaphore.

I think this also makes it more clear that there should be only one thread doing any work at a time.

Reply:
I think we have to spawn the execution for every response with your suggestion. But I might miss something here.

Reply:
Yes, it will take another execution for every response. I think that is not a problem?

Thinking more about this, I think using the `AsyncIOProcessor` would be a good fit for ensuring that the `MultiFileSender` works correctly. It ensures that all requests are executed exactly once and only on one thread at a time (i.e., sequentially). The `Item` passed to it would be the completed seqNo to mark processed. It would still require all transport responses to spawn to the generic thread pool, which then calls `AsyncIOProcessor.put`. Each job would simply mark the seqNo done and check if more requests can be sent now. Since it all runs sequentially, there can be no races. Is that an option?